### PR TITLE
PicoSystem bleepy audio

### PR DESCRIPTION
**Pull request by Gadgetoid**
* Add Pico PWM audio - uses dithered audio output + PIO magic intended for driving a speaker- basically inaudible on PicoSystem
* Add bleepy basic PWM audio - uses GPIO 11's PWM slice to output the first `SQUARE` wave enabled channel to the piezo- quiet but beepy!
* Fix a bug in the audio engine where decay could leave a >0 volume and sustain==0 never turned the envelope generator off.
* Fix some indentation in Pico CMakeLists

### CODE REVIEW

**Comment by Daft-Freak** [comment]
Hmm, is the `audio_pwm` stuff still useful with no users? (The `pio_sm_set_clkdiv` bit needs an `#if OVERCLOCK_250` around it if it's staying...)

**Comment by Gadgetoid** [comment]
> Hmm, is the `audio_pwm` stuff still useful with no users? (The `pio_sm_set_clkdiv` bit needs an `#if OVERCLOCK_250` around it if it's staying...)

I hope we'll find a use for it one day :laughing:

**Comment by Gadgetoid** [comment]
Right, moved `break` as per your demo (nice!) and add a guard around the PIO clock divider.

**Comment by Daft-Freak** [comment]
> I hope we'll find a use for it one day laughing

PicoSystem+ with improved audio :smile:

(I did try it on the VGA board, but it's hardcoded to mono output and i2s is better)

File: 32blit-pico/CMakeLists.txt [hunk 1]
```diff
@@ -45,7 +45,10 @@ if(${PICO_BOARD} STREQUAL "vgaboard")
 	)
 elseif(${PICO_BOARD} STREQUAL "pimoroni_picosystem")
 	message("Using picosystem...")
+	target_link_libraries(BlitHalPico INTERFACE pico_audio_pwm)
```
Comment (Daft-Freak) [inline 32blit-pico/CMakeLists.txt:48]: Don't need this in the end.

File: 32blit-pico/audio.cpp [hunk 2]
```diff
@@ -3,6 +3,25 @@
...
 #include "pico/audio_i2s.h"
 #define HAVE_AUDIO
+#define AUDIO_SAMPLE_FREQ 44100
+#endif
+
+#ifdef AUDIO_PWM
+#include "pico/audio_pwm.h"
+#include "hardware/pio.h"
+#define HAVE_AUDIO
+#define AUDIO_SAMPLE_FREQ 441
```
Comment (Daft-Freak) [inline 32blit-pico/audio.cpp:13]: Uh, 22050? (Though `audio_pwm` ignores you and does ~22050 anyway...)

File: 32blit-pico/audio.cpp [hunk 3]
```diff
@@ -42,24 +71,84 @@ void init_audio() {
...
+      .dma_channel = 1,
+      .pio_sm = 1,
+    },
+    .pattern = 3,
+  };
+  output_format = audio_pwm_setup(&audio_format, -1, &audio_pwm_config);
+  if (!output_format) {
+      panic("PicoAudio: Unable to open audio device.\n");
+  }
+#if OVERCLOCK_250
```
Comment (Daft-Freak) [inline 32blit-pico/audio.cpp:89]: Think this'll need `#include "config.h"`

File: 32blit-pico/audio.cpp (hunk 2 above)
Comment (Gadgetoid) [reply 32blit-pico/audio.cpp]: No idea what happened here :laughing:

File: 32blit-pico/audio.cpp (hunk 3 above)
Comment (Gadgetoid) [reply 32blit-pico/audio.cpp:90]: Oof, done! Thank you.

**Comment by Gadgetoid** [comment]
I'm sneaking a new PicoSystem-specific hardware test in here...

**Comment by Daft-Freak** [comment]
FIFO fix somehow ended up in there despite being merged already?

Maybe the pico-specific hardware test should have a less rainbow-y background, doing a per-pixel `hsv_to_rgba` is pretty slow there.

**Comment by Gadgetoid** [comment]
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA!!!!!!!!!!!!!!!!!

File: 32blit-pico/CMakeLists.txt [hunk 4]
```diff
@@ -45,15 +45,26 @@ if(${PICO_BOARD} STREQUAL "vgaboard")
...
+			INPUT_GPIO
+		)
+	else()
+		target_compile_definitions(BlitHalPico INTERFACE
+			AUDIO_BEEP_PIN=11
+			AUDIO_BEEP
+			DISPLAY_ST7789
+			INPUT_GPIO
+		)
+	endif()
```
Comment (Daft-Freak) [inline 32blit-pico/CMakeLists.txt:64]:
```cmake
	target_compile_definitions(BlitHalPico INTERFACE
		DISPLAY_ST7789
		INPUT_GPIO
	)

	if(AUDIO_PWM)
		target_link_libraries(BlitHalPico INTERFACE pico_audio_pwm)
		target_compile_definitions(BlitHalPico INTERFACE
			PICO_AUDIO_PWM_MONO_PIN=11
			PICO_AUDIO_PWM_PIO=1
			AUDIO_PWM
		)
	else()
		target_compile_definitions(BlitHalPico INTERFACE
			AUDIO_BEEP_PIN=11
			AUDIO_BEEP
		)
	endif()
```

Maybe? Doesn't make it any shorter, but uh... the duplication is annoying

(I have no idea how that file ended up using tabs when everything else is 2 spaces...)

Comment (Gadgetoid) [reply 32blit-pico/CMakeLists.txt:64]: Agreed! We could even define the pins in all cases and just swap AUDIO_PWM/AUDIO_BEEP.

The `PICO_AUDIO_PWM_MONO_PIN` and `AUDIO_BEEP_PIN` should probably come from here somehow: 

https://github.com/raspberrypi/pico-sdk/blob/bfcbefafc5d2a210551a4d9d80b4303d4ae0adf7/src/boards/include/boards/pimoroni_picosystem.h#L52

Comment (Gadgetoid) [reply 32blit-pico/CMakeLists.txt:64]: (Though there appears to be no good way to actually accomplish that)

**Pull request merged**